File: demos/pax-britannica/pax-britannica-android/assets/data/scripts/debris.lua
local v2 = require 'dokidoki.v2'

local SPEED = 0.2
local LIFETIME = 8
local FADE_TIME = 2

local random_direction = v2.unit(math.random() * 2 * math.pi)
local random_scale = math.random() * 0.75 + 0.5
local random_speed = math.random() + 0.5
local random_opacity = math.random() * 0.25 + 0.4

local since_alive = 0

function update()
  since_alive = since_alive + 1/60

  self.transform.scale_x = random_scale
  self.transform.scale_y = random_scale
  
  self.transform.facing = random_direction
  
  self.transform.pos = self.transform.pos + v2.rotate(random_direction, math.sin(since_alive) * 0.5) * SPEED * random_speed
  
  if since_alive < FADE_TIME then
    self.sprite.color[4] = since_alive / FADE_TIME * random_opacity
  else
    self.sprite.color[4] = math.min(1 - (since_alive - LIFETIME + FADE_TIME) / FADE_TIME, 1) * random_opacity
  end
  if since_alive > LIFETIME then self.dead = true end
end

File: demos/pax-britannica/pax-britannica/src/de/swagner/paxbritannica/background/Debris.java
package de.swagner.paxbritannica.background;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;

import de.swagner.paxbritannica.Resources;

public class Debris extends Sprite {

	private float SPEED = 5.0f;
	private float LIFETIME = MathUtils.random(8, 12);
	private float FADE_TIME = 2;

	private float random_direction = MathUtils.random(-360, 360);
	private float random_scale = MathUtils.random() * 0.75f + 0.5f;
	private float random_speed = (MathUtils.random() * 2f) - 1f;
	private float random_opacity = MathUtils.random() * 0.35f + 0.6f;

	private Vector2 position = new Vector2();
	private Vector2 facing = new Vector2(1, 0);

	public boolean alive = true;

	private float since_alive = 0;

	private float delta;

	public Debris(Vector2 position) {
		super();
		this.position = position;
		this.setPosition(position.x, position.y);

		this.facing.rotate(random_direction);
		this.setScale(random_scale, random_scale);

		switch (MathUtils.random(0, 2)) {
		case 0:
			this.set(Resources.getInstance().debrisSmall);
			break;
		case 1:
			this.set(Resources.getInstance().debrisMed);
			break;
		default:
			this.set(Resources.getInstance().debrisLarge);
			break;
		}
	}

	@Override
	public void draw(SpriteBatch spriteBatch) {
		super.draw(spriteBatch);

		delta = Math.min(0.06f, Gdx.graphics.getDeltaTime());

		since_alive += delta;

		facing.rotate((SPEED + random_speed) * delta).nor();
		position.add(facing.scl((SPEED + random_speed) * delta));
		this.setPosition(position.x, position.y);

		if (since_alive < FADE_TIME) {
			super.setColor(1, 1, 1, Math.min((since_alive / FADE_TIME) * random_opacity, random_opacity));
		} else {
			this.setColor(1, 1, 1, Math.min(1 - (since_alive - LIFETIME + FADE_TIME) / FADE_TIME, 1) * random_opacity);
		}
		if (since_alive > LIFETIME) {
			alive = false;
			this.setColor(1, 1, 1, 0);
		}
	}

	public void reset() {
		SPEED = 5.0f;
		LIFETIME = MathUtils.random(8, 12);
		FADE_TIME = 2;

		random_direction = MathUtils.random(-360, 360);
		random_scale = MathUtils.random() * 0.75f + 0.5f;
		random_speed = (MathUtils.random() * 2f) - 1f;
		random_opacity = MathUtils.random() * 0.35f + 0.6f;

		alive = true;
		since_alive = 0;

		this.position = new Vector2(MathUtils.random(-100, 800), MathUtils.random(-100, 400));
		facing = new Vector2(1, 0);

		this.setPosition(position.x, position.y);

		this.facing.rotate(random_direction);
		this.setScale(random_scale, random_scale);
	}
}
